src/network/network.cpp

```diff
@@ -542,6 +555,11 @@ NetworkAddress ParseConnectionString(const std::string &connection_string, uint1
  */
 std::unique_ptr<ServerAddress> ParseGameConnectionString(const std::string &connection_string, uint16 default_port, CompanyID *company)
 {
+	if (!connection_string.empty() && connection_string[0] == '+') {
+		std::string_view ip = ParseCompanyFromConnectionString(connection_string, company);
+		return std::make_unique<ServerAddressJoinKey>(std::string(ip.substr(1)));
```

Comment:
Given I'm seeing the `+` getting added in many places, and it getting removed here, I am wondering why we can't just send it with a `+` over the network. Or, at least once we received it immediately add the `+` to it and only remove it when we send it again.
It would also mean that that ServerDirectAddress and ServerJoinKey classes are basically identical in functionality; it's a type plus a string, so that might remove the need for the specific sub classes and associated casts, and other complications like unique_ptr.

Reply:
I didn't want to send `+` over the network, as this is purely a client-only way method to represent join-keys. The benefit is that it means we can change it to what-ever when-ever we like.

But yeah, the constant adding and removing is a bit annoying; I did try to add/remove it on network level only, but that also wasn't the prettiest code. I will give it another look soon.

Reply:
Did some heavy rewriting on this part, not only with the `+`, but how `ServerAddress` works in general. I think it is a lot cleaner now :)